### PR TITLE
#96: Add Navigation Block

### PR DESCRIPTION
**Pull request by tanvu-epi**

### CODE REVIEW

**Comment by davidknipe** [comment]
This doesn't work in preview mode. Suggest that we implement a method as follows:

`private bool IsPreviewController(BreadcrumbBlock currentBlock)
{
    if (RequestSegmentContext.CurrentContextMode == ContextMode.Edit ||
        RequestSegmentContext.CurrentContextMode == ContextMode.Preview)
    {
        //Retrieve the Request's content stack and get the 2nd element without disturbing the stack.
        var contentStack = Request?.RequestContext.HttpContext.Items[ContentContext.ContentContextKey] as Stack<ContentContext.ContentPropertiesStack>;

        if (contentStack?.Count == 1 &&
            contentStack.Peek().ContentLink.ID == (currentBlock as IContent)?.ContentLink.ID &&
            Request.RawUrl.Contains(",," + (currentBlock as IContent).ContentLink.ID))
        {
            return true;
        }
    }

    return false;
}
`

Then if we are preview mode the build a default model that looks something like this: 

Home / Parent item / Sub item

So the editor can preview what the block can/will look like. Also suggest doing this on the navigation block too.

**Comment by davidknipe** [comment]
NavigationBlock doesn't respect "Display in navigation" into account

**Comment by davidknipe** [comment]
If heading isn't specified please use the current page name as this makes sense as the block should always be aware of the page it is placed on

**Comment by davidknipe** [comment]
Can we allow editors to decide the separator for the breadcrumb? We could use a selection factory as follows (along with associated css) to allow then to select the separator style:

```
public class BreadcrumbSeparatorSelectionFactory : ISelectionFactory
{
    public IEnumerable<ISelectItem> GetSelections(ExtendedMetadata metadata)
    {
        var separators = new List<SelectItem>
        {
            new SelectItem() {Text = "> Single arrow", Value = "bread-crumb-single-arrow"},
            new SelectItem() {Text = "/ Forward slash", Value = "bread-crumb-forward-slash"},
            new SelectItem() {Text = @"\ Backward slash", Value = "bread-crumb-backward-slash"},
            new SelectItem() {Text = "» Double arrow", Value = "bread-crumb-double-arrow"},
            new SelectItem() {Text = "| Pipe", Value = "bread-crumb-pipe"},
            new SelectItem() {Text = ": Pipe", Value = "bread-crumb-colon"}
        };


        return separators;
    }
}
```

File: src/Foundation/Features/Blocks/BreadcrumbBlockController.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+        public BreadcrumbBlockController(IContentLoader contentLoader, IPageRouteHelper pageRouteHelper)
+        {
+            _contentLoader = contentLoader;
+            _pageRouteHelper = pageRouteHelper;
+        }
+        public override ActionResult Index(BreadcrumbBlock currentBlock)
+        {
+            var destination = currentBlock.DestinationPage as ContentReference;
+            if (currentBlock.DestinationPage == null)
```
Comment (lunchin) [inline src/Foundation/Features/Blocks/BreadcrumbBlockController.cs]: ContentReference.IsNullOrEmpty

File: src/Foundation/Features/Blocks/NavigationBlockController.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        public NavigationBlockController(IContentLoader contentLoader, IPageRouteHelper pageRouteHelper)
+        {
+            _contentLoader = contentLoader;
+            _pageRouteHelper = pageRouteHelper;
+        }
+
+        public override ActionResult Index(NavigationBlock currentBlock)
+        {
+            var rootNavigation = currentBlock.RootPage as ContentReference;
+            if (currentBlock.RootPage == null)
```
Comment (lunchin) [inline src/Foundation/Features/Blocks/NavigationBlockController.cs]: Contentreference.IsNullOrEmpty

**Comment by tanvu-epi** [comment]
> This doesn't work in preview mode. Suggest that we implement a method as follows:
> 
> ```
> private bool IsPreviewController(BreadcrumbBlock currentBlock)
> {
>     if (RequestSegmentContext.CurrentContextMode == ContextMode.Edit ||
>         RequestSegmentContext.CurrentContextMode == ContextMode.Preview)
>     {
>         //Retrieve the Request's content stack and get the 2nd element without disturbing the stack.
>         var contentStack = Request?.RequestContext.HttpContext.Items[ContentContext.ContentContextKey] as Stack<ContentContext.ContentPropertiesStack>;
> 
>         if (contentStack?.Count == 1 &&
>             contentStack.Peek().ContentLink.ID == (currentBlock as IContent)?.ContentLink.ID &&
>             Request.RawUrl.Contains(",," + (currentBlock as IContent).ContentLink.ID))
>         {
>             return true;
>         }
>     }
> 
>     return false;
> }
> ```
> 
> Then if we are preview mode the build a default model that looks something like this:
> 
> Home / Parent item / Sub item
> 
> So the editor can preview what the block can/will look like. Also suggest doing this on the navigation block too.

image

Sorry, I don't understand your mean yet, I see Preview still work for Breadcrumb and Navigation block. Please describe more clearly, thanks.

**Comment by davidknipe** [comment]
> > This doesn't work in preview mode. Suggest that we implement a method as follows:
> > ```
> > private bool IsPreviewController(BreadcrumbBlock currentBlock)
> > {
> >     if (RequestSegmentContext.CurrentContextMode == ContextMode.Edit ||
> >         RequestSegmentContext.CurrentContextMode == ContextMode.Preview)
> >     {
> >         //Retrieve the Request's content stack and get the 2nd element without disturbing the stack.
> >         var contentStack = Request?.RequestContext.HttpContext.Items[ContentContext.ContentContextKey] as Stack<ContentContext.ContentPropertiesStack>;
> > 
> >         if (contentStack?.Count == 1 &&
> >             contentStack.Peek().ContentLink.ID == (currentBlock as IContent)?.ContentLink.ID &&
> >             Request.RawUrl.Contains(",," + (currentBlock as IContent).ContentLink.ID))
> >         {
> >             return true;
> >         }
> >     }
> > 
> >     return false;
> > }
> > ```
> > 
> > 
> > Then if we are preview mode the build a default model that looks something like this:
> > Home / Parent item / Sub item
> > So the editor can preview what the block can/will look like. Also suggest doing this on the navigation block too.
> 
> image
> 
> Sorry, I don't understand your mean yet, I see Preview still work for Breadcrumb and Navigation block. Please describe more clearly, thanks.

When creating a new block, without specifying the target page this happens:

https://www.dropbox.com/s/5dmhm2jm1ira7c6/breadcrumberror.mp4?dl=0 

As a general rule editors should not be able to do anything that will intentionally generate a 500

**Comment by tanvu-epi** [comment]
> > > This doesn't work in preview mode. Suggest that we implement a method as follows:
> > > ```
> > > private bool IsPreviewController(BreadcrumbBlock currentBlock)
> > > {
> > >     if (RequestSegmentContext.CurrentContextMode == ContextMode.Edit ||
> > >         RequestSegmentContext.CurrentContextMode == ContextMode.Preview)
> > >     {
> > >         //Retrieve the Request's content stack and get the 2nd element without disturbing the stack.
> > >         var contentStack = Request?.RequestContext.HttpContext.Items[ContentContext.ContentContextKey] as Stack<ContentContext.ContentPropertiesStack>;
> > > 
> > >         if (contentStack?.Count == 1 &&
> > >             contentStack.Peek().ContentLink.ID == (currentBlock as IContent)?.ContentLink.ID &&
> > >             Request.RawUrl.Contains(",," + (currentBlock as IContent).ContentLink.ID))
> > >         {
> > >             return true;
> > >         }
> > >     }
> > > 
> > >     return false;
> > > }
> > > ```
> > > 
> > > 
> > > Then if we are preview mode the build a default model that looks something like this:
> > > Home / Parent item / Sub item
> > > So the editor can preview what the block can/will look like. Also suggest doing this on the navigation block too.
> > 
> > 
> > image
> > Sorry, I don't understand your mean yet, I see Preview still work for Breadcrumb and Navigation block. Please describe more clearly, thanks.
> 
> When creating a new block, without specifying the target page this happens:
> 
> https://www.dropbox.com/s/5dmhm2jm1ira7c6/breadcrumberror.mp4?dl=0
> 
> As a general rule editors should not be able to do anything that will intentionally generate a 500

Fixed in the latest commit, branch Feature/96-add-navigation-block

**Comment by lunchin** [comment]
@davidknipe are you okay with this now?

**Pull request merged**